### ISSUE TITLE
Propagate APM tracing configuration to sub-processes via the control protocol

### ISSUE THREAD

**Issue by joshdover** [elastic/elastic-agent#2612]
The primary goal of this issue is to enabling trace collection on Integrations Server on Cloud by default to send to our internal APM backends. This tracing configuration will be specified in the `elastic-agent.yml` file on the server side and will not be part of the agent policy retrieved by Fleet.

See the initial discussion in https://github.com/elastic/elastic-agent/issues/1969

### Requirements

- When tracing is configured in Agent's local configuration, but absent in the Fleet agent policy, the local tracing configuration should take precedence
- APM tracing configuration must be propagated to sub-processes via the control protocol
	- We should not use environment variables since this will not always be possible in some runtimes
- elastic-agent-client should be updated to support this configuration
- We need to prevent any sensitive information from leaking via APM tracing data
	- We should audit existing traces
	- We should propose / consider any additional protections that need to be enabled to avoid leaking sensitive data
	- Areas to focus on: requests to Fleet Server, handling of the agent policy (local and remote from Fleet), gRPC control protocol messages to sub-processes

**Comment by joshdover** [comment]
APM configuration will be used for more than just sending traces, as applications like Fleet Server and APM Server also use this backend for metrics (and likely one day, logs too).

I think we should change the plan here slightly to start sending down an OTLP endpoint to sub-processes, which can then be used for all of these use cases. In the future, this endpoint may point to something local to the Agent or shipper as a gateway before forwarding to the backend.

As part of this, we'd probably want to change the agent yaml configuration to allow specifying an OTLP destination, but I think we can do that afterwards, since the existing APM server URL for tracing is also an OTLP endpoint.

**Comment by cmacknz** [comment]
> As part of this, we'd probably want to change the agent yaml configuration to allow specifying an OTLP destination, but I think we can do that afterwards, since the existing APM server URL for tracing is also an OTLP endpoint.

We need to define how we model this in the agent policy, considering that there is no OTLP output today and if there were we would likely relying on a dedicated OTLP monitoring output to do this.

The shortest path would be to put something dedicated in the `agent.monitoring` section:

https://github.com/elastic/elastic-agent/blob/7cb529542759528e8a055771b113a51fd1cb5713/elastic-agent.reference.yml#L130-L136

If we were just going to add `traces`, I would propose we add a `traces` configuration object that takes the OTLP destination as a sub-object like `traces.otlp` or similar. This is easy to do because our tracer always supports OTLP output.

This wouldn't work for metrics though because the `agent.monitoring.metrics` key isn't an object, and every agent sub-process won't support OTLP metrics output initially. We could add an `agent.monitoring.otlp` section with the destination configuration and then have separate flags for each data source like `agent.monitoring.otlp.metrics: true` to use it for specific data types. 

For a concrete example we could end up with something like:


```yaml
agent.monitoring:
  # Enable monitoring.
  enabled: true
  # Collect metrics from processes
  metrics: true
  # Collect logs from processes
  logs: true
  # Collect traces from processes
  traces: true
  # OpenTelemetry protocol configuration
  otlp:
    # OTLP protocol host or IP
    host: xxxx
    # TLS configuration (we use the ssl key for this everywhere else already)
    ssl: xxxx
    # Use this otlp destination for traces
    traces: true
    # Use this otlp destination for metrics (where supported).
    metrics: true
```

The only thing I don't like about this is that setting `agent.monitoring.otlp.metrics: true` might enable otlp metrics, or it might not, because not all inputs would support it and there's no way for a user to know besides turning it on and seeing what happens.

**Comment by joshdover** [comment]
Thanks for giving this some additional thought.

> The only thing I don't like about this is that setting `agent.monitoring.otlp.metrics: true` might enable otlp metrics, or it might not, because not all inputs would support it and there's no way for a user to know besides turning it on and seeing what happens.

I would expect that (eventually) the metrics that Agent currently scrapes from input processes also go through OTLP when `agent.monitoring.otlp.metrics` is enabled. We can do this because we know the data model of our input metrics today and can transform those into OTLP metric records in a straightforward way. Until then, you are right that we need to keep this option experimental.

Longer-term we will probably want to support a dedicated OTLP output type, which agent monitoring should be able to reference. How does that overlap with this? I wasn't able to find an example of how to set the monitoring output to a different destination.

For the scope of this ticket though, I don't think it needs to matter too much. I think in any case the sub-processes need to receive a specific OTLP endpoint for monitoring data, and how it's configured in elastic-agent.yml is something we can iterate on, as long as we keep it experimental for now.

**Comment by pierrehilbert** [comment]
@joshdover / @cmacknz I think this OTLP topic should be addressed as a next step to iterate and avoid to be too ambitious in the next sprint.
For sure we should have in mind the big picture and those information will be useful to @pchila but let's create a followup issue. WDYT?

**Comment by jlind23** [comment]
@pierrehilbert what do you mean? Starting with OLTP traces within the monitoring section and work on the metrics part in a follow up Pr?

**Comment by pierrehilbert** [comment]
What I mean is to focus on traces and we will see later for the other parts.
Can we agree on that?

**Comment by cmacknz** [comment]
We can use traces as the primary implementation and exclude anything else, since our APM tracing library supports OTLP out of the box.

I apparently missed that we have an existing APM configuration section in the agent policy that Josh mentioned in https://github.com/elastic/elastic-agent/issues/2612#issuecomment-1628416820. This is because it isn't in our reference configuration but it is in the code.

https://github.com/elastic/elastic-agent/blob/3fefe42ed7967351a4b35c62069aaffcf8d4ba49/internal/pkg/core/monitoring/config/config.go#L22

https://github.com/elastic/elastic-agent/blob/3fefe42ed7967351a4b35c62069aaffcf8d4ba49/internal/pkg/core/monitoring/config/config.go#L67-L82

Given that this already exists, I think we can simplify this further:

1. Let's not modify the agent policy at all. Since our existing APM endpoint is also an OTLP endpoint we can keep using it experimentally for this purpose. 
2. When we add support for passing this APM configuration to sub-processes, it should be modelled as an OTLP endpoint. This can be as simple as renaming APM to OTLP everywhere, but we should set the expectation for sub-processes that this is a generic OTLP endpoint and not an Elastic APM endpoint specifically to future proof it. We should review https://www.elastic.co/guide/en/apm/guide/current/open-telemetry-direct.html for the configuration needed to forward OTLP traces to APM server. We need to make sure our configuration supports both the HTTP and gRPC formats.

In summary, we should put all of our effort into generalizing the control protocol messages and leave the agent policy alone. The control protocol is much harder to change in the future since it affects teams outside of the Elastic Agent team.

**Comment by joshdover** [comment]
@cmacknz this matches my thinking as well. I'd prefer to avoid modifying the agent policy configuration at all for now and keep this as undocumented feature while we mature the solution and standardize all of Agent's internal monitoring on OTLP.

One snag is that elastic-apm-go doesn't yet support output as OTLP, which I didn't realize. So if we pass down an OTLP endpoint now, I think we'll need to make a short-term assumption that it actually supports APM intake v2 _and_ OTLP until we can change our applications to adopt the OTel SDK (this is the plan for APM Server already).

I'm personally ok with making that assumption for now as long as it really is short-term. Otherwise, I'd prefer we expose the endpoint as explicitly being an APM endpoint so we can more smoothly phase this out with a pure OTLP endpoint later.

**Comment by cmacknz** [comment]
> One snag is that elastic-apm-go doesn't yet support output as OTLP, which I didn't realize. So if we pass down an OTLP endpoint now, I think we'll need to make a short-term assumption that it actually supports APM intake v2 and OTLP until we can change our applications to adopt the OTel SDK (this is the plan for APM Server already).

> I'm personally ok with making that assumption for now as long as it really is short-term. Otherwise, I'd prefer we expose the endpoint as explicitly being an APM endpoint so we can more smoothly phase this out with a pure OTLP endpoint later.

This makes me think it is actually better to add explicit support for the APM endpoint today, and just define the messages in a way that allows us to also provide an OTLP endpoint in the future. We could just add a monitoring message type to the protocol that can contain the various endpoints we support as they become available.

I would much rather have the protocol reflect the current reality, than assume that certain things will happen in the future and have to keep explaining that the assumptions in the protocol are wrong for months :).

**Comment by joshdover** [comment]
We got feedback from the ESS team that they need to be able to supply the global_labels option as a map/key-value pairs in the elastic-agent.yml file instead of a comma-separated string. I spoke with @pchila about this and it seems feasible to change now.

**Comment by pchila** [comment]
@dmathieu 
Leaving here a summary of what has been implemented in Elastic Agent in PR elastic/elastic-agent#3223:
- The APM elastic agent configuration in `elastic-agent.yml` looks like this (the field names under `apm` have the same meaning and usage as the regular [APM configuration](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html) :
  ```yaml
  agent.monitoring:
    traces: true
    apm:
      hosts:
        - <apm host url>
      environment: <apm environment>
      secret_token: <redacted>
      api_key: <redacted>
      global_labels:
        k1: v1
        k2: v2
      tls:
        skip_verify: true
        server_certificate: <path to the server certificate>
        server_ca: <path to the server CA>
  ```
- The configuration is only available in the configuration file (Fleet does not support this at the moment), so:
  - in case of standalone agent the configuration is reloaded from file in case of changes while the agent is running
  - in case of managed agent, the configuration is read once at startup and then added to every policy change coming from Fleet: in this case changes to APM configuration require a restart of agent to be picked up
- APM propagation to components requires `agent.monitoring.traces` to be set to `true` 
- At the moment the agent supports only Elastic APM configuration but since want to support OTLP protocol the APM configuration has a dedicated field for Elastic and we will put support for other protocols side-by-side (see [APMConfig protobuf definition](https://github.com/elastic/elastic-agent-client/blob/5c7929a9889af5047137fabcb8f16ea38653ab97/elastic-agent-client.proto#L188-L208) )
- APM configuration is passed to components into each unit: it's the component responsibility to apply it by reloading/re-instantiating its own APM instrumentation. See a small example in the [fake component used to test the feature](https://github.com/elastic/elastic-agent/pull/3223/files#diff-8d0667c5f6ec618d526400f021043af11de5f94afe2357ecb69aca58757e2c28R181-R196)

**Comment by cmacknz** [comment]
@pchila might be worth adding the contents of that comment to https://github.com/elastic/elastic-agent/tree/main/docs. You can create a tracing.md to summarize the current state of things.

**Comment by pchila** [comment]
I thought about it but it seems that this is an internal functionality, if I put the description in the docs doesn't it become kinda generally available? Maybe we should add a disclaimer saying that this is an internal feature and may change at any moment ?

**Comment by simitt** [comment]
>that this is an internal feature and may change at any moment ?

Is this feature going to be released as GA or beta? Even if used internal, the configuration should not change afterwards, as it would be breaking downstream products using it, such as the apm-server.

**Comment by dmathieu** [comment]
Also, is there a private repository with design docs somewhere for this doc to sit?
Side note: this repository is public. So technically, the doc is already out there (though less visible than if it were a file).

**Comment by cmacknz** [comment]
We should document this in the documentation in the agent repository docs folder for internal use cases (as mentioned the code is already here and public), and it will be omitted from the official documentation and release notes since we aren't intending to support this for users initially.
